xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+            }
+          }
+          if (!principalNames.isEmpty()) {
+            return principalNames;
+          }
+        }
+        return Collections.singleton(cert.getSubjectDN().getName());
+      } catch (SSLPeerUnverifiedException | CertificateParsingException ex) {
+        return null;
+      }
+    }
+
+    private Metadata getHeader() {
+      return metadata;
+    }
+
+    private InetAddress getDestinationIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private InetAddress getSourceIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private int getDestinationPort() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? -1 : ((InetSocketAddress) addr).getPort();
+    }
+  }
+
+  abstract static class Matcher {
+    protected Matcher() {
+    }
+
+    abstract boolean matches(EvaluateArgs args);
+  }
+
+  /** Matches when any of the matcher matches. */
+  static final class OrMatcher extends Matcher {
+    private final List<? extends Matcher> anyMatch;
+
+    OrMatcher(List<? extends Matcher> matchers) {
+      checkNotNull(matchers, "matchers");
+      this.anyMatch = Collections.unmodifiableList(matchers);
+    }
+
+    static OrMatcher create(Matcher...matchers) {
+      return new OrMatcher(Arrays.asList(matchers));
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      for (Matcher m : anyMatch) {
+        if (m != null && m.matches(args)) {
+          return true;
+        }
+      }
+      return false;
+    }
+  }
+
+  /** Matches when all of the matchers match. */
+  static final class AndMatcher extends Matcher {
+    private final List<? extends Matcher> allMatch;
+
+    AndMatcher(List<? extends Matcher> matchers) {
+      checkNotNull(matchers, "matchers");
+      this.allMatch = Collections.unmodifiableList(matchers);
+    }
+
+    static AndMatcher create(Matcher...matchers) {
+      return new AndMatcher(Arrays.asList(matchers));
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      for (Matcher m : allMatch) {
+        if (m == null || !m.matches(args)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+
+  /** Always true matcher.*/
+  static final class AlwaysTrueMatcher extends Matcher {
+    static AlwaysTrueMatcher INSTANCE = new AlwaysTrueMatcher();
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return true;
+    }
+  }
+
+  /** Negate matcher.*/
+  static final class InvertMatcher extends Matcher {
+    private final Matcher toInvertMatcher;
+
+    InvertMatcher(Matcher matcher) {
+      this.toInvertMatcher = checkNotNull(matcher);
```

Comment:
nit: Include the second argument `"matcher"`. It really doesn't matter much here, but it helps a _lot_ other times it means you don't need accurate line numbers, so it is just best to do it all the time.